<p style="font-family: Arial; font-size:3em;color:black;"> Session 4 - Lab Exercise 7</p>

1. Kevin Xavier Antony Arul Xavier
2. Vikas Baliyan
3. Ramuel John Batuigas
4. Chirag Ashokkumar Bhatia
5. Jazbi Izhar
6. Sidhartha Sankar Mondal
7. Bharat Sharma

In [133]:
# you can download the original Kyphosis dataset from kaggle at https://www.kaggle.com/abbasit/kyphosis-dataset?select=kyphosis.csv

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [134]:
Kyphosis_df = pd.read_csv('kyphosis.csv')

In [135]:
from sklearn.model_selection import train_test_split

In [136]:
X = Kyphosis_df.drop('Kyphosis',axis=1)
y = Kyphosis_df['Kyphosis']

In [137]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [138]:
from sklearn.tree import DecisionTreeClassifier

In [139]:
# selecting the right "parameters/hyperparameters of ML models can significantly affect their performacnce".
# let's begin exploring those parameters more for a DT model.
'''
DecisionTreeClassifier(
    criterion='gini',
    splitter='best',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features=None,
    random_state=None,
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    min_impurity_split=None,
    class_weight=None,
    presort=False,
)
'''
# Let's consider these three parameters: criterion, splitter, min_samples_split
    # for parameters consider two cases 'gini' and 'entropy'
    # for splitter consider two cases 'best' and 'random'
    # for min_samples_split consider two cases 2 and 5
    # explain and dicuss your findings

"\nDecisionTreeClassifier(\n    criterion='gini',\n    splitter='best',\n    max_depth=None,\n    min_samples_split=2,\n    min_samples_leaf=1,\n    min_weight_fraction_leaf=0.0,\n    max_features=None,\n    random_state=None,\n    max_leaf_nodes=None,\n    min_impurity_decrease=0.0,\n    min_impurity_split=None,\n    class_weight=None,\n    presort=False,\n)\n"

In [140]:
from sklearn.metrics import accuracy_score

In [141]:
# Model 1

dtc = DecisionTreeClassifier(criterion = 'gini',splitter = 'best', min_samples_split = 2)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.68

In [142]:
# Model 2

dtc = DecisionTreeClassifier(criterion = 'entropy',splitter = 'best', min_samples_split = 2)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.8

In [143]:
# Model 3

dtc = DecisionTreeClassifier(criterion = 'gini',splitter = 'random', min_samples_split = 2)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.68

In [144]:
# Model 4

dtc = DecisionTreeClassifier(criterion = 'entropy',splitter = 'random', min_samples_split = 2)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.72

In [145]:
# Model 5

dtc = DecisionTreeClassifier(criterion = 'gini',splitter = 'best', min_samples_split = 5)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.76

In [146]:
# Model 6

dtc = DecisionTreeClassifier(criterion = 'entropy',splitter = 'best', min_samples_split = 5)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.8

In [147]:
# Model 7

dtc = DecisionTreeClassifier(criterion = 'gini',splitter = 'random', min_samples_split = 5)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.68

In [149]:
# Model 8

dtc = DecisionTreeClassifier(criterion = 'entropy',splitter = 'random', min_samples_split = 5)
dtc.fit(X_train,y_train)
y_pred = dtc.predict(X_test)
accuracy_score(y_test,y_pred)

0.88

### Observations

1. Model 8 works the best with 88% Accuracy result
2. Model 8 Parameters are criterion = 'entropy', splitter = 'random' and min_sample_split = 5
3. In Model 8 entropy gives the best results 